harfbuzz_overflow_resolution.md

```diff
@@ -0,0 +1,157 @@
+   
+The harfbuzz subsetting library
+[includes a repacking algorithm](https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-repacker.hh)
+which is used to resolve offset overflows that are present in the subsetted tables it produces. This
+document provides a deep dive into how the harfbuzz repacking algorithm works.
+
+# Foundations
+
+There's three key pieces to the harfbuzz approach:
+
+*  Graph representation: a table's internal structure is represented by a lightweight graph
+   representation where each subtable is a node and each offset forms an edge. The actual subtable
+   bytes are not stored in each node. Instead we just store start and end memory addresses for each.
+   This lightweight representation can be quickly modified as the repacking algorithm iterates.
+
+*  Topological sorting algorithm: an algorithm which given a graph produces a topological sorting
+   of the nodes.
+   
+*  Overflow check: given a graph and a topological sorting checks if there will be any overflows
+   in any of the offsets. If there are overflows returns a list of (parent, child) tuples that
+   will overflow. Since the graph has information on the size of each subtable it's straightforward
+   to calculate the final position of each subtable and then check if any offsets to it will
+   overflow.
+   
+# High Level Algorithm
+
+```
+var graph
+graph.topological_sort()
+
+while (overflows = graph.will_overflow()):
+  for overflow in overflows:
+    attempt_offset_resolution_strategy(overflow, graph)
+  graph.topological_sort()
+
+```
+
+The actual code for this processing loop can be found [here](https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-repacker.hh#L682).
+
+# Topological Sorting Algorithms
+
+The harfbuzz repacker uses two different algorithms for topological sorting:
+*  [Kahn's Algorithm](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm)
+*  Sorting by shortest distance
+
+Kahn's algorithm is approximately twice as fast as the shortest distance sort so that is attempted
+first (only on the first topological sort). If it fails to eliminate overflows then shortest distance
+sort will be used for all subsequent topological sorting operations.
+   
+## Shortest Distance Sort
+
+This algorithm orders the nodes based on total distance to each node. Nodes with a shorter distance
+are ordered first.
+
+The "weight" of an edge is the sum of the size of the sub-table being pointed to plus 2^16 for a 16 bit
+offset and 2^32 for a 32 bit offset.
+
+The distance of a node is the sum of all weights along the shortest path from the root to that node
+plus a priority modifier (used to change where nodes are placed by moving increasing or
+decreasing the effective distance). Ties between nodes with the same distance are broken based
+on the order of the offset in the sub table bytes.
+
+The shortest distance to each node is determined using Djikstra's algorithm. Then the topological
```

Comment:
although relatively well known it wouldn't hurt to link to Djikstra's since you did Kahn's above